store/multiversion/store.go

```diff
@@ -0,0 +1,122 @@
+}
+
+type Store struct {
+	mtx sync.RWMutex
+	// map that stores the key -> MultiVersionValue mapping for accessing from a given key
+	multiVersionMap map[string]MultiVersionValue
+	// TODO: do we need to add something here to persist readsets for later validation
+	// TODO: we need to support iterators as well similar to how cachekv does it
+	// TODO: do we need secondary indexing on index -> keys - this way if we need to abort we can replace those keys with ESTIMATE values? - maybe this just means storing writeset
+}
+
+func NewMultiVersionStore() *Store {
+	return &Store{
+		multiVersionMap: make(map[string]MultiVersionValue),
+	}
+}
+
+// GetLatest implements MultiVersionStore.
+func (s *Store) GetLatest(key []byte) (value MultiVersionValueItem) {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+
+	keyString := conv.UnsafeBytesToStr(key)
+	// if the key doesn't exist in the overall map, return nil
+	if _, ok := s.multiVersionMap[keyString]; !ok {
+		return nil
+	}
+	val, found := s.multiVersionMap[keyString].GetLatest()
+	if !found {
+		return nil // this shouldn't be possible
+	}
+	return val
+}
+
+// GetLatestBeforeIndex implements MultiVersionStore.
+func (s *Store) GetLatestBeforeIndex(index int, key []byte) (value MultiVersionValueItem) {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+
+	keyString := conv.UnsafeBytesToStr(key)
+	// if the key doesn't exist in the overall map, return nil
+	if _, ok := s.multiVersionMap[keyString]; !ok {
+		return nil
+	}
+	val, found := s.multiVersionMap[keyString].GetLatestBeforeIndex(index)
+	// otherwise, we may have found a value for that key, but its not written before the index passed in
+	if !found {
+		return nil
+	}
+	// found a value prior to the passed in index, return that value (could be estimate OR deleted, but it is a definitive value)
+	return val
+}
+
+// Has implements MultiVersionStore. It checks if the key exists in the multiversion store at or before the specified index.
+func (s *Store) Has(index int, key []byte) bool {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+
+	keyString := conv.UnsafeBytesToStr(key)
+	if _, ok := s.multiVersionMap[keyString]; !ok {
+		return false // this is okay because the caller of this will THEN need to access the parent store to verify that the key doesnt exist there
+	}
+	_, found := s.multiVersionMap[keyString].GetLatestBeforeIndex(index)
+	return found
+}
+
+// This function will try to intialize the multiversion item if it doesn't exist for a key specified by byte array
+// NOTE: this should be used within an acquired mutex lock
+func (s *Store) tryInitMultiVersionItem(keyString string) {
+	if _, ok := s.multiVersionMap[keyString]; !ok {
+		multiVersionValue := NewMultiVersionItem()
+		s.multiVersionMap[keyString] = multiVersionValue
+	}
+}
+
+// Set implements MultiVersionStore.
+func (s *Store) Set(index int, incarnation int, key []byte, value []byte) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	keyString := conv.UnsafeBytesToStr(key)
```

Comment:
i'd prefer us to just use string casting here. If key bytes are mutated somehow the underlying map would be corrupted and would be difficult to debug

Reply:
what do you mean by mutated somehow? This is how cachekv currently implements the cache that maps between byte key and vals, so i would think there isnt too much risk in doing it this way?

Reply:
modified to use string cast for safety btw